packages/ui/src/components/multi-select.tsx

```diff
@@ -0,0 +1,158 @@
+import { ChangeEvent, ReactNode, useEffect, useMemo, useRef, useState } from 'react'
+
+import {
+  DropdownMenu,
+  DropdownMenuContent,
+  DropdownMenuItem,
+  DropdownMenuSeparator,
+  DropdownMenuTrigger,
+  Icon,
+  ScrollArea,
+  SearchBox
+} from '@/components'
+import { cn } from '@utils/cn'
+import { TFunction } from 'i18next'
+import { debounce } from 'lodash-es'
+
+export type MultiSelectOptionType<T = unknown> = T & {
+  id: string | number
+  label: string
+}
+
+export interface MultiSelectProps<T = unknown> {
+  selectedItems: MultiSelectOptionType<Partial<T>>[]
+  t: TFunction
+  placeholder: string
+  handleChange: (data: MultiSelectOptionType<Partial<T>>) => void
+  options: MultiSelectOptionType<T>[]
+  searchValue?: string
+  handleChangeSearchValue?: (data: string) => void
+  customOptionElem?: (data: MultiSelectOptionType<T>) => ReactNode
+}
+
+export const MultiSelect = <T = unknown,>({
+  selectedItems,
+  t,
+  placeholder,
+  handleChange,
+  options,
+  searchValue = '',
+  handleChangeSearchValue,
+  customOptionElem
+}: MultiSelectProps<T>) => {
+  const [search, setSearch] = useState('')
+
+  const debouncedChangeSearchRef = useRef(debounce((value: string) => handleChangeSearchValue?.(value), 300))
+
+  useEffect(() => setSearch(searchValue ?? ''), [searchValue])
+
+  useEffect(() => {
+    const debouncedChangeSearch = debouncedChangeSearchRef.current
+
+    return () => {
+      debouncedChangeSearch.cancel()
+    }
+  }, [])
```

Comment:
i see this pattern repeated multiple times.. of creating refs for debounce functions which are then cancelled in a useEffect.
why is this needed?